File: backend/models/student.py
from datetime import datetime, date, timezone
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel, Relationship
from typing_extensions import Annotated, Literal


class Gender:
    """Gender constants"""
    MALE = "male"
    FEMALE = "female"
    OTHER = "other"


class StudentStatus:
    """Student status constants"""
    ACTIVE = "active"
    GRADUATED = "graduated"
    WITHDRAWN = "withdrawn"
    TRANSFERRED = "transferred"
    SUSPENDED = "suspended"


class Class(SQLModel, table=True):
    """Class model for organizing students"""

    __tablename__ = "classes"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    school_id: UUID = Field(foreign_key="schools.id")
    name: Annotated[str, Field(max_length=100)]
    level: Annotated[str, Field(max_length=50)]
    class_teacher_id: Optional[UUID] = Field(default=None, foreign_key="users.id")
    capacity: int = Field(default=40)
    academic_year: Annotated[str, Field(max_length=20)]
    is_active: bool = Field(default=True)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class Student(SQLModel, table=True):
    """Student model"""

    __tablename__ = "students"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    school_id: UUID = Field(foreign_key="schools.id")
    admission_number: Annotated[str, Field(max_length=50, unique=True, index=True)]
    first_name: Annotated[str, Field(max_length=100)]
    last_name: Annotated[str, Field(max_length=100)]
    other_names: Optional[str] = Field(default=None, max_length=100)
    date_of_birth: date
    gender: Literal["male", "female", "other"]
    guardian_name: Annotated[str, Field(max_length=200)]
    guardian_phone: Annotated[str, Field(max_length=20)]
    guardian_email: Optional[str] = Field(default=None, max_length=255)
    guardian_relationship: Annotated[str, Field(max_length=50)]
    address: Annotated[str, Field(max_length=500)]
    class_id: Optional[UUID] = Field(default=None, foreign_key="classes.id")
    status: Literal["active", "graduated", "withdrawn", "transferred", "suspended"] = Field(
        default="active"
    )
    photo_url: Optional[str] = Field(default=None, max_length=500)
    enrollment_date: date
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: Optional[datetime] = Field(default=None)

File: src/types/student.ts
export type StudentStatus = 'active' | 'graduated' | 'withdrawn' | 'transferred' | 'suspended';
export type Gender = 'male' | 'female' | 'other';

export interface Student {
  id: string;
  admission_number: string;
  first_name: string;
  last_name: string;
  other_names: string | null;
  date_of_birth: string;
  gender: Gender;
  guardian_name: string;
  guardian_phone: string;
  guardian_email: string | null;
  guardian_relationship: string;
  address: string;
  class_id: string | null;
  status: StudentStatus;
  photo_url: string | null;
  enrollment_date: string;
  created_at: string;
  updated_at: string | null;
}

export interface CreateStudent {
  admission_number: string;
  first_name: string;
  last_name: string;
  other_names?: string;
  date_of_birth: string;
  gender: Gender;
  guardian_name: string;
  guardian_phone: string;
  guardian_email?: string;
  guardian_relationship?: string;
  address: string;
  class_id?: string;
  photo_url?: string;
}

export interface UpdateStudent {
  first_name?: string;
  last_name?: string;
  other_names?: string;
  guardian_name?: string;
  guardian_phone?: string;
  guardian_email?: string;
  address?: string;
  class_id?: string;
  photo_url?: string;
  status?: StudentStatus;
}

export interface Class {
  id: string;
  name: string;
  level: string;
  class_teacher_id: string | null;
  capacity: number | null;
  academic_year: string;
  is_active: boolean;
}
